libraries/AP_BattMonitor/AP_BattMonitor_JBDCAN.cpp

```diff
@@ -0,0 +1,124 @@
+#include "AP_BattMonitor_JBDCAN.h"
+#include <GCS_MAVLink/GCS.h>
+#include <cstring>  // memset, memcpy
+#include <cstdio>
+extern const AP_HAL::HAL& hal;
+
+// Constructor: lưu con trỏ đến state và param
+AP_BattMonitor_JBDCAN::AP_BattMonitor_JBDCAN(AP_BattMonitor& mon,
+                                             AP_BattMonitor::BattMonitor_State& mon_state,
+                                             AP_BattMonitor_Params& params) :
+    AP_BattMonitor_Backend(mon, mon_state, params),
+    _can_interface_index(0)  // Mặc định sử dụng CAN1 (index 0)
+
+
+{}
+
+// Lấy CAN interface (mặc định là CAN1 = hal.can[0])
+void AP_BattMonitor_JBDCAN::init() {
+    _iface = hal.can[1];
+    if (_iface == nullptr) {
+        GCS_SEND_TEXT(MAV_SEVERITY_ERROR, "JBDCAN: CAN interface not available");
+        return;
+    }
+    // Khởi tạo CAN: 500kbps, chế độ Normal
+    if (!_iface->init(500000, AP_HAL::CANIface::NormalMode)) {
+        GCS_SEND_TEXT(MAV_SEVERITY_ERROR, "JBDCAN: Failed to initialize CAN interface");
+        return;
+    }
+    if (!_iface->register_frame_callback(
+        FUNCTOR_BIND_MEMBER(&AP_BattMonitor_JBDCAN::handle_frame_callback, void, uint8_t, const AP_HAL::CANFrame &, AP_HAL::CANIface::CanIOFlags),
+        _callback_id)) {
+        GCS_SEND_TEXT(MAV_SEVERITY_ERROR, "JBDCAN: Failed to register frame callback");
+        return;
+    }
+    _last_update_us = AP_HAL::micros64();
+    GCS_SEND_TEXT(MAV_SEVERITY_INFO, "JBDCAN: Interface ready");
+    
+
+}
+
+
+
+
+void AP_BattMonitor_JBDCAN::read()
+{
+    if (_iface == nullptr) return;
+
+    uint64_t now = AP_HAL::micros();
+    if (now - _last_update_us < 200000) return;
+
+    uint16_t id = 0x100;
+    AP_HAL::CANFrame rtr_frame;
+    rtr_frame.id = id | AP_HAL::CANFrame::FlagRTR;
+    rtr_frame.dlc = 0;
+    rtr_frame.setCanFD(false);
+
+    int16_t ret = _iface->send(rtr_frame, now + 100000, 0);
+    if (ret > 0) {
+        _last_update_us = now;
+    } else if (ret < 0) {
+        GCS_SEND_TEXT(MAV_SEVERITY_WARNING, "JBDCAN: Failed to send RTR frame");
+    }
+}
+
+void AP_BattMonitor_JBDCAN::handle_frame_callback(uint8_t iface_num, const AP_HAL::CANFrame &frame, AP_HAL::CANIface::CanIOFlags flags)
+{
+    // Chỉ xử lý frame có ID phù hợp (response từ JBD BMS)
+    if ((frame.id & 0xFFFF) == 0x100) {
+        handle_frame(frame);
+    }
+}
+
+    
+void AP_BattMonitor_JBDCAN::handle_frame(const AP_HAL::CANFrame &frame)
+{   
+    
+    if (frame.dlc < 6 || frame.dlc > 8) {
+        GCS_SEND_TEXT(MAV_SEVERITY_WARNING, "JBDCAN: Frame length invalid");
+        return;
+    }
+
+    const uint8_t *data = frame.data;
+
+    uint16_t raw_voltage = (data[0] << 8) | data[1];
+    float voltage = raw_voltage * 0.01f;
+
+    int16_t raw_current = (data[2] << 8) | data[3];
+    float current = raw_current * 0.01f;
+
+    uint16_t raw_capacity = (data[4] << 8) | data[5];
```

Comment:
consider casting to a structure and extracting eg `float voltage` directly from that structure (inc le16toh and whatnot if required when taking from the structure)